### PR TITLE
refactor: Update the list of files to be stored and delivered

### PR DESCRIPTION
**Pull request by ivadym**
### This PR:

Updates the list of files that are generated and stored by BALSAMIC.

- Removed from `.hk` file: 
  - `fastp` rule:
    - `_1.fp.fastq.gz`
    - `_2.fp.fastq.gz`
    - `_fastp.json`
    - `_fastp.html`
  - `bcftools_filter_tnscope_tumor_only` rule:
    - `.tnscope_isec.all.filtered.vcf.gz`
  - `bcftools_filter_tnhaplotyper_tumor_normal` and  `bcftools_filter_tnhaplotyper_tumor_only` rules:
    - `.tnhaplotyper.all.filtered.vcf.gz`
    - `.tnhaplotyper.all.filtered.pass.vcf.gz`

- Made temporal:
  - `fastp` rule:
    - `_1.fp.fastq.gz`
    - `_2.fp.fastq.gz`
  - `fastp_umi` rule:
    - `_1.umi_optimized.fastq.gz`
    - `_2.umi_optimized.fastq.gz`

Hermes PR: https://github.com/Clinical-Genomics/hermes/issues/44

### Review and tests: 
- [ ] Tests pass
- [ ] Code review
- [ ] New code is executed and covered by tests, and test approve

### CODE REVIEW

File: BALSAMIC/constants/workflow_rules.py [hunk 1]
```diff
@@ -113,31 +113,26 @@
...
-    "bcftools_filter_svdb",
-    "bcftools_intersect_tumor_only",
-    "bcftools_filter_TNscope_umi_tumor_only",
-    "genmod_score_vardict",
     "mergeBam_tumor",
     "mergeBam_normal",
     "mergeBam_tumor_umiconsensus",
     "mergeBam_normal_umiconsensus",
-    "cnvkit_paired",
+    "vep_somatic_sv",
```
Comment (khurrammaqbool) [inline BALSAMIC/constants/workflow_rules.py:122]: This should be replaced with svdb_merge_tumor_normal and svdb_merge_tumor_only. We decided to include raw VCFs and annotated filtered pass VCFs- vep_somatic_sv rule does not generate raw VCFs.

File: BALSAMIC/snakemake_rules/annotation/vep.rule [hunk 2]
```diff
@@ -56,7 +56,7 @@ rule vep_somatic_sv:
     vcf = vcf_dir + "SV.somatic.{case_name}.svdb.vcf.gz",
     header = vcf_dir + "SV.somatic.{case_name}.svdb.sample_name_map",
   output:
-    vcf_all = temp(vep_dir + "SV.somatic.{case_name}.svdb.all.vcf.gz"),
+    vcf_all_svdb = vep_dir + "SV.somatic.{case_name}.svdb.all.vcf.gz",
```
Comment (khurrammaqbool) [inline BALSAMIC/snakemake_rules/annotation/vep.rule:59]: This file should be replaced with the one in vcf folder from the rules mentioned above.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule [hunk 3]
```diff
@@ -126,16 +126,19 @@ rule ascat_tumor_normal:
...
-        ascat_plots= expand(
-            vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat." + "{output_suffix}" + ".png",
-            output_suffix=["ascatprofile", "rawprofile", "ASPCF", "tumor", "germline", "sunrise"]
-        ),
+        plot_ascat_profile = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.ascatprofile.png",
+        plot_raw_profile = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.rawprofile.png",
+        plot_aspcf = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.ASPCF.png",
+        plot_tumor = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.tumor.png",
+        plot_germline = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.germline.png",
+        plot_sunrise = vcf_dir + "CNV.somatic." + config["analysis"]["case_id"] + ".ascat.sunrise.png",
```
Comment (khurrammaqbool) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:134]: The final pdf contains all these plots with the table. Isn't it better to deliver the pdf file instead?

Comment (ashwini06) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:134]: Indeed, good to deliver the final pdf.

File: BALSAMIC/snakemake_rules/quality_control/fastp.rule [hunk 4]
```diff
@@ -82,7 +82,6 @@ rule fastp:
     singularity: 
         Path(singularity_image, config["bioinfo_tools"].get("fastp") + ".sif").as_posix() 
     params:
-        housekeeper_id = {"id": "{sample}", "tags": "quality-trimmed-fastq"},
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/quality_control/fastp.rule:85]: Can you also mark the `.umi_optimised.fastq.gz` and `.fp.fastq.gz` as temporary. Related to https://github.com/Clinical-Genomics/BALSAMIC/issues/911

File: BALSAMIC/constants/workflow_rules.py (hunk 1 above)
Comment (ashwini06) [reply BALSAMIC/constants/workflow_rules.py:122]: https://github.com/Clinical-Genomics/BALSAMIC/blob/ad13bdfaf5cdb8b754bae35c40d6c1f4a112f27d/BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule#L216

https://github.com/Clinical-Genomics/BALSAMIC/blob/ad13bdfaf5cdb8b754bae35c40d6c1f4a112f27d/BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule#L107

Inclusion of hk tags for these rules

File: BALSAMIC/constants/workflow_rules.py [hunk 5]
```diff
@@ -113,31 +113,26 @@
...
+    "vep_somatic_snv",
```
Comment (ashwini06) [inline BALSAMIC/constants/workflow_rules.py:123]: Inline to above changes, for somatic snvs it should follow the raww vcf deliverables from https://github.com/Clinical-Genomics/BALSAMIC/blob/ad13bdfaf5cdb8b754bae35c40d6c1f4a112f27d/BALSAMIC/snakemake_rules/variant_calling/somatic_tumor_normal.rule#L49 (panel), umi_panel https://github.com/Clinical-Genomics/BALSAMIC/blob/ad13bdfaf5cdb8b754bae35c40d6c1f4a112f27d/BALSAMIC/snakemake_rules/umi/sentieon_varcall_tnscope_tn.rule#L7,   https://github.com/Clinical-Genomics/BALSAMIC/blob/ad13bdfaf5cdb8b754bae35c40d6c1f4a112f27d/BALSAMIC/snakemake_rules/variant_calling/sentieon_tn_varcall.rule#L156 (wgs) . Same for T-only

**Review by ashwini06 — CHANGES_REQUESTED** [review]
@ivadym : Can you also please link the related Hermes PR to this PR.

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule (hunk 3 above)
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_normal.rule:134]: @ashwini06, @khurrammaqbool:

Adding a HK tag to these rules does not mean that this files will be delivered. The excel that I've shared with you contains that list.

The problem is that I added `ascat_tumor_normal` to the delivery rules lisT. However, `rule.py` does not support a `namedlist` as an input (generated by `expand(...)`), so it was crashing here:
 https://github.com/Clinical-Genomics/BALSAMIC/blob/8244f837388404d236611349d9eac4cb094290d1/BALSAMIC/utils/rule.py#L207

**Comment by ivadym** [comment]
Linking the Hermes PR: https://github.com/Clinical-Genomics/hermes/issues/44

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule [hunk 6]
```diff
@@ -113,13 +113,14 @@ rule svdb_merge_tumor_only:
...
     output:
-        svdb_vcf = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".svdb.vcf.gz",
+        vcf_svdb = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".svdb.vcf.gz",
         namemap = vcf_dir + "SV.somatic." + config["analysis"]["case_id"] + ".svdb.sample_name_map",
     benchmark:
         Path(benchmark_dir, 'svdb_merge_tumor_only_' + config["analysis"]["case_id"] + ".tsv")
     singularity:
         Path(singularity_image, config["bioinfo_tools"].get("svdb") + ".sif").as_posix()
     params:
+        housekeeper_id = {"id": config["analysis"]["case_id"], "tags": "research"},
```
Comment (ashwini06) [inline BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule:123]:
```suggestion
        housekeeper_id = {"id": config["analysis"]["case_id"], "tags": "research"},
```
Is the tag here should be `research` or `clinical` ? Is there any documentation  to look up about these Hermes tags to describe what is `research` and `clinical`?

**Comment by ashwini06** [comment]
@ivadym : Is it possible to fix the tests for few of these missing lines? https://app.codecov.io/gh/Clinical-Genomics/BALSAMIC/compare/915/changes

File: BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule (hunk 6 above)
Comment (ivadym) [reply BALSAMIC/snakemake_rules/variant_calling/somatic_sv_tumor_only.rule:123]: The documentation in Hermes is a big vague: https://github.com/Clinical-Genomics/hermes/blob/14808b8ffa6b722ef1c603680e42f1eebe9237f9/cg_hermes/config/tags.py.

But I believe MIP follows this structure of defining the raw/annotated vcfs as `research` and the filtered ones as `clinical`.

File: BALSAMIC/constants/workflow_rules.py [hunk 7]
```diff
@@ -113,31 +113,31 @@
...
-    "bcftools_intersect_tumor_only",
-    "bcftools_filter_TNscope_umi_tumor_only",
-    "genmod_score_vardict",
     "mergeBam_tumor",
     "mergeBam_normal",
     "mergeBam_tumor_umiconsensus",
     "mergeBam_normal_umiconsensus",
-    "cnvkit_paired",
+    "vep_germline",
+    "tmb_calculation",
```
[the remark this answers is not in the document]
Comment (ivadym) [reply BALSAMIC/constants/workflow_rules.py]: Done here: [d354b32](https://github.com/Clinical-Genomics/BALSAMIC/pull/915/commits/d354b328065991181d415007bf003175549b6316)

**Comment by ivadym** [comment]
Quality control results:
```
/qc/sample_tumor_normal_wgs_metrics_deliverables.yaml
/qc/multiqc_report.html
/qc/multiqc_data/multiqc_data.json
```

Analysis specific results:

- TN-WGS
```
        /vep/SNV.germline.tumor.dnascope.vcf.gz
        /vep/SNV.germline.normal.dnascope.vcf.gz
        /vep/SV.germline.tumor.manta_germline.vcf.gz
        /vep/SV.germline.normal.manta_germline.vcf.gz
        /vcf/SNV.somatic.sample_tumor_normal_wgs.tnscope.vcf.gz
        /vcf/SNV.somatic.sample_tumor_normal_wgs.tnhaplotyper.vcf.gz
        /vcf/SV.somatic.sample_tumor_normal_wgs.svdb.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal_wgs.tnscope.all.filtered.pass.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal_wgs.tnhaplotyper.all.filtered.pass.vcf.gz
        /vep/SV.somatic.sample_tumor_normal_wgs.svdb.all.filtered.pass.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal_wgs.tnscope.balsamic_stat
        /vep/SNV.somatic.sample_tumor_normal_wgs.tnhaplotyper.balsamic_stat
        /vcf/CNV.somatic.sample_tumor_normal_wgs.ascat.output.pdf
        /vcf/CNV.somatic.sample_tumor_normal_wgs.ascat.copynumber.txt.gz
```

- TN-TGA
```
        /vep/SNV.germline.tumor.haplotypecaller.vcf.gz
        /vep/SNV.germline.normal.haplotypecaller.vcf.gz
        /vep/SNV.germline.tumor.dnascope.vcf.gz
        /vep/SNV.germline.normal.dnascope.vcf.gz
        /vep/SV.germline.tumor.manta_germline.vcf.gz
        /vep/SV.germline.normal.manta_germline.vcf.gz
        /vcf/SNV.somatic.sample_tumor_normal.TNscope_umi.vcf.gz
        /vcf/SNV.somatic.sample_tumor_normal.vardict.vcf.gz
        /vcf/SNV.somatic.sample_tumor_normal.tnhaplotyper.vcf.gz
        /vcf/SV.somatic.sample_tumor_normal.svdb.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal.TNscope_umi.all.filtered.pass.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal.vardict.all.filtered.pass.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal.tnhaplotyper.all.filtered.pass.vcf.gz
        /vep/SV.somatic.sample_tumor_normal.svdb.all.filtered.pass.vcf.gz
        /vep/SNV.somatic.sample_tumor_normal.vardict.balsamic_stat
        /vep/SNV.somatic.sample_tumor_normal.tnhaplotyper.balsamic_stat
        /vep/SNV.somatic.sample_tumor_normal.TNscope_umi.balsamic_stat
        /cnv/tumor.merged.cns
        /cnv/tumor.merged-diagram.pdf
        /cnv/tumor.merged-scatter.pdf
        /cnv/sample_tumor_normal.gene_metrics
        /vcf/CNV.somatic.sample_tumor_normal.cnvkit.vcf2cytosure.cgh
        /vep/SNV.somatic.sample_tumor_normal.vardict.all.filtered.pass.ranked.vcf.gz
        /qc/umi_qc/concatenated_normal_XXXXXX_R.umi.mean_family_depth
        /qc/umi_qc/concatenated_tumor_XXXXXX_R.umi.mean_family_depth
        /qc/umi_qc/sample_tumor_normal.TNscope_umi.AFtable.txt
```

**Comment by ivadym** [comment]
Files and tags are parsed and generated as expected:

- WGS_TN

Screen Shot 2022-04-28 at 20 43 23

- TGA_TN

Screen Shot 2022-04-28 at 21 17 05

**Review by ashwini06 — APPROVED** [review]
Looks good 🎇

**Comment by ashwini06** [comment]
> Files and tags are parsed and generated as expected:
> 
> * WGS_TN
> 
> Screen Shot 2022-04-28 at 20 43 23
> 
> * TGA_TN
> 
> Screen Shot 2022-04-28 at 21 17 05

@ivadym What does the warning message "Cannot deliver step (rule) ..." refers to?

**Comment by ivadym** [comment]
> @ivadym What does the warning message "Cannot deliver step (rule) ..." refers to?

It means that the rule it tries to deliver is not among the list of all the rules executed during the analysis.

If you notice we are going through all the rules that we are going to deliver. However, among all of them we can find some that are specific to the analysis type. Therefore, for `TN_WGS`, for example, you cannot find the rules associated with the UMIs.

https://github.com/Clinical-Genomics/BALSAMIC/blob/8244f837388404d236611349d9eac4cb094290d1/BALSAMIC/workflows/balsamic.smk#L299-L304

**Review by khurrammaqbool — COMMENTED** [review]
👍 . 
As we discussed, let's refactor to decouple delivery after this.

**Review by khurrammaqbool — APPROVED** [review]

**Pull request merged**